File: Tutorials/Codes/Modern_Control_Course/Chapter21/Lesson1/Chapter21_Lesson1.m
% Chapter21_Lesson1.m
% Finite transmission zeros from the Rosenbrock system matrix.
% MATLAB functions used:
%   ss, tzero   : Control System Toolbox, if available
%   rank, svd   : core numerical linear algebra
%   syms, det   : Symbolic Math Toolbox, if available
%
% Simulink section:
%   Creates a simple State-Space block model if Simulink is installed.

clear; clc;

A = [0 1; -2 -3];
B = [0; 1];
C = [4 1];
D = 0;

fprintf('System: G(s) = (s + 4)/(s^2 + 3s + 2)\n');

% Control System Toolbox route
if exist('ss', 'file') == 2 && exist('tzero', 'file') == 2
    sys = ss(A, B, C, D);
    z = tzero(sys);
    disp('Transmission zeros from tzero(sys):');
    disp(z);
else
    disp('Control System Toolbox tzero not found; using Rosenbrock rank check.');
end

% Rosenbrock rank check at z = -4
z0 = -4;
Rz = [z0*eye(size(A,1)) - A, -B; C, D];
disp('R(-4) =');
disp(Rz);
fprintf('rank R(-4) = %d\n', rank(Rz));
fprintf('normal rank for this SISO square case = %d\n', size(A,1) + size(B,2));

% Symbolic determinant route
if exist('syms', 'file') == 2
    syms s
    R = [s*eye(size(A,1)) - A, -B; C, D];
    zeroPolynomial = factor(det(R));
    disp('det R(s) =');
    disp(zeroPolynomial);
    disp('roots(det R(s)) =');
    disp(double(solve(zeroPolynomial == 0, s)));
else
    disp('Symbolic Math Toolbox not found; skipping symbolic determinant.');
end

% Verify zero-output exponential trajectory
x0 = [1; -4];
u0 = 6;
stateResidual = z0*x0 - A*x0 - B*u0;
outputResidual = C*x0 + D*u0;
disp('state residual z*x0 - A*x0 - B*u0 =');
disp(stateResidual);
disp('output residual C*x0 + D*u0 =');
disp(outputResidual);

% Optional Simulink model generation
if exist('simulink', 'file') == 2
    modelName = 'Chapter21_Lesson1_Simulink';
    if bdIsLoaded(modelName)
        close_system(modelName, 0);
    end
    new_system(modelName);
    open_system(modelName);

    add_block('simulink/Sources/Step', [modelName '/StepInput'], ...
        'Position', [60 80 100 110]);
    add_block('simulink/Continuous/State-Space', [modelName '/StateSpacePlant'], ...
        'A', mat2str(A), 'B', mat2str(B), 'C', mat2str(C), 'D', mat2str(D), ...
        'Position', [170 70 300 120]);
    add_block('simulink/Sinks/Scope', [modelName '/OutputScope'], ...
        'Position', [380 75 430 115]);

    add_line(modelName, 'StepInput/1', 'StateSpacePlant/1');
    add_line(modelName, 'StateSpacePlant/1', 'OutputScope/1');

    save_system(modelName);
    disp(['Simulink model saved as ', modelName, '.slx']);
else
    disp('Simulink not found; skipping model generation.');
end

File: Tutorials/Codes/Modern_Control_Course/Chapter21/Lesson1/Chapter21_Lesson1.py
# Chapter21_Lesson1.py
# Finite transmission zeros from the Rosenbrock system matrix.
# Required packages: numpy, scipy, sympy
#
# Model:
#   x_dot = A x + B u
#   y     = C x + D u
#
# A complex number z is a finite transmission zero if the Rosenbrock matrix
#   R(z) = [[zI - A, -B],
#           [C,       D]]
# loses rank relative to its normal rank.

import numpy as np
import scipy.linalg as la
import sympy as sp


def rosenbrock_matrix(A, B, C, D, s):
    """Return the Rosenbrock system matrix R(s)."""
    A = np.asarray(A, dtype=complex)
    B = np.asarray(B, dtype=complex)
    C = np.asarray(C, dtype=complex)
    D = np.asarray(D, dtype=complex)

    n = A.shape[0]
    top = np.hstack((s * np.eye(n, dtype=complex) - A, -B))
    bottom = np.hstack((C, D))
    return np.vstack((top, bottom))


def numerical_rank(M, tol=1e-9):
    """SVD-based numerical rank."""
    singular_values = la.svd(M, compute_uv=False)
    return int(np.sum(singular_values > tol))


def siso_zeros_by_rosenbrock_determinant(A, B, C, D):
    """
    For a square SISO Rosenbrock matrix, det(R(s)) gives the finite invariant
    zero polynomial up to a nonzero scalar factor when the realization is regular.
    """
    s = sp.symbols("s")
    A_sp = sp.Matrix(A)
    B_sp = sp.Matrix(B)
    C_sp = sp.Matrix(C)
    D_sp = sp.Matrix(D)
    n = A_sp.rows

    R = A_sp.zeros(n + C_sp.rows, n + B_sp.cols)
    R[:n, :n] = s * sp.eye(n) - A_sp
    R[:n, n:] = -B_sp
    R[n:, :n] = C_sp
    R[n:, n:] = D_sp

    zero_polynomial = sp.factor(R.det())
    roots = sp.nroots(zero_polynomial)
    return zero_polynomial, roots, R


def main():
    # Minimal SISO example with transfer function G(s) = (s + 4)/(s^2 + 3s + 2).
    A = np.array([[0.0, 1.0],
                  [-2.0, -3.0]])
    B = np.array([[0.0],
                  [1.0]])
    C = np.array([[4.0, 1.0]])
    D = np.array([[0.0]])

    zero_polynomial, zeros, R_symbolic = siso_zeros_by_rosenbrock_determinant(A, B, C, D)
    print("Symbolic Rosenbrock matrix R(s):")
    print(R_symbolic)
    print("\ndet(R(s)) =", zero_polynomial)
    print("Transmission zeros =", zeros)

    z = complex(-4.0, 0.0)
    Rz = rosenbrock_matrix(A, B, C, D, z)
    print("\nR(-4) =")
    print(np.real_if_close(Rz))
    print("rank R(-4) =", numerical_rank(Rz))
    print("normal rank for this SISO square example =", A.shape[0] + B.shape[1])

    # Verify the zero-output exponential trajectory:
    # x(t) = exp(z t) x0, u(t) = exp(z t) u0.
    x0 = np.array([[1.0], [-4.0]])
    u0 = np.array([[6.0]])
    residual_state = (z * x0) - A @ x0 - B @ u0
    residual_output = C @ x0 + D @ u0
    print("\nState residual z*x0 - A*x0 - B*u0 =")
    print(np.real_if_close(residual_state))
    print("Output residual C*x0 + D*u0 =")
    print(np.real_if_close(residual_output))


if __name__ == "__main__":
    main()
